### PR TITLE
Feature/increase fleet long poll timeout

### LINKED ISSUE #2257: Increase Fleet long poll default to ~30 minutes
We've been doing scale testing over the past few months using a ~30 minute long poll duration (rather than current default of 5m) and we are seeing much better results for very large clusters.

We're now ready to make this the default setting for Fleet Server and Agent. These changes can happen independently and do not necessarily need to land in the same release, though it would be preferred. The corresponding Fleet Server changes are in tracked in:
- https://github.com/elastic/fleet-server/issues/2337

There is some additional complexity to changing this on the Agent side, as we currently have an issue where Agent will not re-checkin with Fleet Server when it's health status changes. If we update the long polling interval to 30 minutes, this could result in the agent status in the UI being up to 30 minutes stale, rather than only 5 minutes stale.

To avoid this kind of regression, we need to update Agent to also cancel the current checkin and start a new one when status changes, however we will cap the frequency of this to 5 minutes to avoid any extra load on large Fleets. We will investigate increasing the frequency that Agent updates this further separately from this change, see https://github.com/elastic/elastic-agent/issues/1946.

### Tasks

The client side timeout in Agent should be longer than Fleet Server (28m) or the proxy's timeout (30m 20s). We'll keep a similar buffer here at 5 minutes over what the proxy will timeout at and timeout at 35 minutes from the client.

- [ ] Add the ability cancel a checkin and start a new one when the status changes, with a 5 minute debounce
- [ ] Update default for `fleet.timeout` to 35 minutes: https://github.com/elastic/elastic-agent/blob/c0976970f2317dd79a9056ea1ee2044b72485713/internal/pkg/remote/config.go#L49

### PR DESCRIPTION
**Pull request by pchila**
## What does this PR do?

- Increase the fleet checkin timeout to align it with the new Fleet long poll settings
- Improve logging messages pointing to the troubleshooting guide in case of checkin errors
- Refactor the fleet gateway to perform the checkin asynchronously with the option of cancelling it if a new state of the agent must be sent to fleet.

*Note for reviewers*: you can safely ignore `mock_*_test.go` files since those are generated mocks  used for unit testing and a lot of them have just been moved

## Why is it important?
This PR implements agent-side modifications to improve fleet/agent scalability as described on #2257.
It  should reduce the load on Fleet server side allowing it to manage thousands of agents as verified during scale testing.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [ ] I have added and integration test or an e2E test

## Author's Checklist

- [x] Test with a new 8.8.0 snapshot fleet with the new long polling checkin request (pass `poll_timeout` parameter)

## How to test this PR locally
- Enroll an agent with an 8.8.0-SNAPSHOT fleet server with a policy of your choice (the default one should be enough)
- wait till a checkin is in progress with a healthy agent state
- change the agent state (in this example we remove the executable flag from metricbeat and kill all the running instances of it)
  ```
  chmod -x ./data/elastic-agent-781ccb/components/metricbeat
  killall metricbeat
  
  
  ./elastic-agent status
  State: DEGRADED
  Message: 1 or more components/units in a failed state
  Fleet State: STARTING
  Fleet Message: (no message)
  Components:
    * filestream    (HEALTHY)
                    Healthy: communicating with pid '433570'
    * http/metrics  (FAILED)
                    Failed: execution of component prevented: not executable by owner
    * beat/metrics  (FAILED)
                    Failed: execution of component prevented: not executable by owner
  
  ```
- after an initial debounce of 5 mins, the agent should cancel the checkin in progress and start a new one with the new (in my case `degraded` ) state.
- bring the agent back to health and after another 5 mins debounce you should see that also the second checkin is cancelled and another one is started for the `healthy` state
  ```
  chmod +x ./data/elastic-agent-781ccb/components/metricbeat
  <wait a moment for the agent to restart the processes>
  ./elastic-agent status
  State: HEALTHY
  Message: Running
  Fleet State: STARTING
  Fleet Message: (no message)
  Components:
    * filestream    (HEALTHY)
                    Healthy: communicating with pid '433570'
    * http/metrics  (HEALTHY)
                    Healthy: communicating with pid '435918'
    * beat/metrics  (HEALTHY)
                    Healthy: communicating with pid '435917'
  ```
- if the agent state does not change, the checkin duration can go up to ~30 mins (depending on when fleet writes back a response)

This is what it looks like when the log level is set to debug

```json
{"log.level":"info","@timestamp":"2023-04-19T19:00:52.408+0200","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":160},"message":"Fleet gateway started","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-04-19T19:00:52.408+0200","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":161},"message":"Fleet gateway settings: &{Debounce:5m0s Duration:1s Jitter:500ms Backoff:{Init:1m0s Max:10m0s}}","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-04-19T19:00:52.778+0200","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":169},"message":"FleetGateway calling Checkin API","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-04-19T19:00:52.778+0200","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":238},"message":"starting checkin \"7d4b304c-7304-4089-918d-1771e8fc174e\"","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-04-19T19:00:52.778+0200","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":312},"message":"Checkin started","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-04-19T19:05:52.779+0200","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":262},"message":"Received updated state (Agent state: \"DEGRADED\") when checkin is ongoing past configured debounce. Cancelling previous checkin \"7d4b304c-7304-4089-918d-1771e8fc174e\" and starting a new one.","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-04-19T19:05:52.779+0200","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":280},"message":"Cancelling checkin \"7d4b304c-7304-4089-918d-1771e8fc174e\"","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-04-19T19:05:52.780+0200","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":318},"message":"Ongoing checkin with fleet-server has been explicitly canceled, won't be retried","log":{"source":"elastic-agent"},"error":{"message":"fail to checkin to fleet-server: all hosts failed: 1 error occurred:\n\t* requester 0/1 to host https://fb70f4283adb4a61ab1ac498fd4617b3.fleet.eastus2.azure.qa.cld.elstc.co:9243/ errored: Post \"https://fb70f4283adb4a61ab1ac498fd4617b3.fleet.eastus2.azure.qa.cld.elstc.co:9243/api/fleet/agents/f3819377-b5a2-4f68-9a8e-956c90f65328/checkin?\": context canceled\n\n"},"request_duration_ns":299778262270,"failed_checkins":0,"ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-04-19T19:05:52.780+0200","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":284},"message":"Reaped answer for cancelled checkin \"7d4b304c-7304-4089-918d-1771e8fc174e\": {response:<nil> err:{msg:fail to checkin to fleet-server err:0xc000170140 errType:5 meta:map[uri:/api/fleet/agents/f3819377-b5a2-4f68-9a8e-956c90f65328/checkin]}}","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-04-19T19:05:52.780+0200","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":238},"message":"starting checkin \"3c22a4a9-b877-4e80-b7aa-9f06cad3c153\"","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-04-19T19:05:52.780+0200","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":312},"message":"Checkin started","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-04-19T19:10:52.780+0200","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":262},"message":"Received updated state (Agent state: \"HEALTHY\") when checkin is ongoing past configured debounce. Cancelling previous checkin \"3c22a4a9-b877-4e80-b7aa-9f06cad3c153\" and starting a new one.","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-04-19T19:10:52.780+0200","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":280},"message":"Cancelling checkin \"3c22a4a9-b877-4e80-b7aa-9f06cad3c153\"","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-04-19T19:10:52.780+0200","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":318},"message":"Ongoing checkin with fleet-server has been explicitly canceled, won't be retried","log":{"source":"elastic-agent"},"error":{"message":"fail to checkin to fleet-server: all hosts failed: 1 error occurred:\n\t* requester 0/1 to host https://fb70f4283adb4a61ab1ac498fd4617b3.fleet.eastus2.azure.qa.cld.elstc.co:9243/ errored: Post \"https://fb70f4283adb4a61ab1ac498fd4617b3.fleet.eastus2.azure.qa.cld.elstc.co:9243/api/fleet/agents/f3819377-b5a2-4f68-9a8e-956c90f65328/checkin?\": context canceled\n\n"},"request_duration_ns":299770438043,"failed_checkins":0,"ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-04-19T19:10:52.780+0200","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":284},"message":"Reaped answer for cancelled checkin \"3c22a4a9-b877-4e80-b7aa-9f06cad3c153\": {response:<nil> err:{msg:fail to checkin to fleet-server err:0xc0001701a0 errType:5 meta:map[uri:/api/fleet/agents/f3819377-b5a2-4f68-9a8e-956c90f65328/checkin]}}","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-04-19T19:10:52.780+0200","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":238},"message":"starting checkin \"9edaed78-e4d3-4947-ae0f-47bb72a9f469\"","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2023-04-19T19:10:52.781+0200","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":312},"message":"Checkin started","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
```

## Related issues

-  Closes #2257

### CODE REVIEW

**Comment by joshdover** [comment]
We discussed today that we should add a new option to horde to be able to tell an agent to short-circuit the long poll from the server and start a new one on some configurable interval. This would allow testing of worst-case scenarios like "what if every agent is experiencing 'flapping' health issues and checking in the maximum amount of time".

In theory this would only be "as bad" as our current scalability already is (we can already achieve many of our scale targets with the current behavior), so the floor to this should be the current state.

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (blakerouse) [inline internal/pkg/agent/application/gateway/fleet/fleet_gateway.go:261]: Why is this here?

**Review by blakerouse — APPROVED** [review]
This looks really good. The Fleet Gateway code is rather dirty to begin with and I know you where trying to not change it to much, with the worry that you would affect the behavior. I was able to understand your additions and changes even though its not the cleanest.

There is just one inline comment, that seems like it should be removed. Other than that it looks good to me.

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (pchila) [reply internal/pkg/agent/application/gateway/fleet/fleet_gateway.go:261]: to avoid the linter screaming at me while I was developing it and the variable was unused. good catch!

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (michel-laterman) [inline internal/pkg/agent/application/gateway/fleet/fleet_gateway.go:452]: You're going to need to change the `CheckinRequest` to add a `poll_timeout` string attribute.

**Review by michel-laterman — CHANGES_REQUESTED** [review]
CheckinRequests need a `poll_timeout` duration in the body to poll longer then the default time

**Comment by pchila** [comment]
@michel-laterman the `long_poll` parameter has been added to the checkin request. Could you have another look ?

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (AndersonQ) [inline internal/pkg/agent/application/gateway/fleet/fleet_gateway.go:220]: Did you forget the `go`? otherwise why are you defining and immediately after invoking a anonymous function?
If it to use defer and return?

**Review by AndersonQ — CHANGES_REQUESTED** [review]

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (pchila) [reply internal/pkg/agent/application/gateway/fleet/fleet_gateway.go:214]: That is a debug log that I needed for debugging the agent while developing the async checkin (mostly to understand that we got some non-error answer and we are going to process the actions)

Since we can have more or less the same information by specifiying `DEBUG_AGENT=1` in env I am going to remove the checking response value from here ;)

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (pchila) [reply internal/pkg/agent/application/gateway/fleet/fleet_gateway.go:277]: I wasn't sure if we wanted to see when we are cancelling a checkin with a given state because of an update.
If we are not interested this can be moved to `Debugf`

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (pchila) [reply internal/pkg/agent/application/gateway/fleet/fleet_gateway.go:290]: More debugging logs since there was some funny behavior when draining channels on defer()... I will leave only the log after cancelling as it should be sufficient

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (blakerouse) [reply internal/pkg/agent/application/gateway/fleet/fleet_gateway.go]: To make this cleaner and easier to read it would be better to just move this to its own function.

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (pchila) [reply internal/pkg/agent/application/gateway/fleet/fleet_gateway.go:328]: Downgraded to Info

File: internal/pkg/agent/configuration/fleet.go
Comment (blakerouse) [reply internal/pkg/agent/configuration/fleet.go:32]: +1

**Review by blakerouse — CHANGES_REQUESTED** [review]
Overall this is looking really good. The testing coverage you have added on the gateway client is fantastic. I do see one implementation issue that we should address before this is merged. It is something that I didn't think about when we discussed this originally, but I think its something we need to add.

File: internal/pkg/fleetapi/checkin_cmd.go
Comment (pchila) [reply internal/pkg/fleetapi/checkin_cmd.go:58]: Added more explanation to the `CheckinDuration` comment as that is actually the main reason for the type alias

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (pchila) [reply internal/pkg/agent/application/gateway/fleet/fleet_gateway.go:282]: So, if I understand correctly, you would like to have something like a small delay before triggering a cancel in case the state goes back to what was previously... That is gonna be interesting to shoehorn in the current code structure...

I could start refactoring the debounce right away instead of waiting for #2504 if you like the proposed solution

Comment (blakerouse) [reply internal/pkg/agent/application/gateway/fleet/fleet_gateway.go:282]: Yes.

I am good with the solution proposed in #2504 for the debounce as long as its a wrapper over the subscription and not something that changes the subscription code and messaging (some things need the information instantly without debounce).

Comment (pchila) [reply internal/pkg/agent/application/gateway/fleet/fleet_gateway.go:282]: The idea here is to have 2 debouncers that have an input channel, an output channel and an internal state for the last value received.
If we create a small pipe composed of a one-off 5 mins debouncer and a resettable 30 seconds debouncer we should be fine and the subscription code has no changes

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (pchila) [reply internal/pkg/agent/application/gateway/fleet/fleet_gateway.go]: Done

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (blakerouse) [reply internal/pkg/agent/application/gateway/fleet/fleet_gateway.go:286]: I like the implementation idea.

I would prefer to see it in this PR, because I worry that we will drop it and not get it done. I would rather have it out of the gate instead of waiting.

**Review by michel-laterman — DISMISSED** [review]
lgtm, just have some minor nitpicks
Can you make sure that the new debounce + poll_timeout are in the developer docs as well?

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (pchila) [reply internal/pkg/agent/application/gateway/fleet/fleet_gateway.go:232]: This id associated to a checkin is there to be able to track what happens to checkin requests during the cancellatio/relaunch (since we now may have hanging cancellations). If we have a unique id generated from a span I am happy to use that as an identifier so that we can track it across the network boundary. I'll have a look and see if I can integrate it in this PR or if it takes too long I will create a separate issue for that

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (pchila) [reply internal/pkg/agent/application/gateway/fleet/fleet_gateway.go]: After our discussion yesterday, I added a debouncer which behaves like before but it allows for cleaner code in 781ccb7aee7fe9b7b15dadb870b122f4ae6f46ca

I also tried to define a debouncer which will return a bit of history of the states received during the debounce period in 42aae248551848f7fbfae3ca3d90bc3ae5ae06fd so that we can consolidate a state by discarding transient ones (we need to define what amounts as transient) but even this will not be a perfect solution (it will heavily depend on the heuristics we use).

I am going to try and integrate the AccumulatorDebouncer in the fleet gateway but I am starting to feel that we are trying to solve a different problem with this and we need to discuss what amounts as "good enough" for this PR.

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (pchila) [reply internal/pkg/agent/application/gateway/fleet/fleet_gateway.go:235]: Opened #2517

**Comment by cmacknz** [comment]
Let's add a markdown document explaining how check ins work as of 8.8.0 to https://github.com/elastic/elastic-agent/tree/main/docs along side the explanation of the fleet bootstrap process.

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (pchila) [reply internal/pkg/agent/application/gateway/fleet/fleet_gateway.go]: As agreed with @cmacknz, @blakerouse, the behaviour implemented with a simple debounce is no worse than what we already have.
Debouncing of transient agent states will not be included in this PR but it will be a followup issue.

**Comment by pchila** [comment]
Testing this PR using an elastic cloud deployment revealed a timeout after 30 minutes:
```json
{"log.level":"warn","@timestamp":"2023-04-20T17:31:40.029+0200","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":338},"message":"Possible transient error during check in with fleet-server, retrying","log":{"source":"elastic-agent"},"error":{"message":"status code: 0, fleet-server returned an error: , message: Post \"https://172.23.200.15:18242/api/fleet/agents/3a1b2c7f-07c8-4156-9dc6-676955664a5b/checkin\": net/http: request canceled (Client.Timeout exceeded while awaiting headers)"},"request_duration_ns":1800668254794,"failed_checkins":1,"retry_after_ns":78484350481,"ecs.version":"1.6.0"}
```

Setting the Long poll duration to 30 minutes works as Fleet answers in ~ 28 minutes.
Still unclear where this timeout is coming from, need to be investigated more in depth

File: changelog/fragments/1681919580-increase-fleet-long-poll-timeout.yaml
Comment (cmacknz) [inline changelog/fragments/1681919580-increase-fleet-long-poll-timeout.yaml:14]: You want this to be a human readable description, the changelog tool doesn't make this clear.

I would state what the timeout has increased to, why, and what effects it might have. In particular this changelog entry should call out that people using non-Elastic proxies may have to adjust their connection timeouts.

**Review by blakerouse — DISMISSED** [review]
Change looks good, looks ready to merge. I really like the debug logging you added that will be very helpful if we come across issues with this new logic!

File: docs/fleet-long-polls.md
Comment (AndersonQ) [inline docs/fleet-long-polls.md:26]:
```suggestion
- creates a subcontext from the main Fleet Gateway context and uses it to subscribe to state updates from Coordinator (abstracted away as a StateFetcher interface)
```

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (AndersonQ) [inline internal/pkg/agent/application/gateway/fleet/fleet_gateway.go:83]: typo?
```suggestion
// checkinState represents the data we send about agent state in fleet checkin requests.
```

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (AndersonQ) [inline internal/pkg/agent/application/gateway/fleet/fleet_gateway.go:202]: [Suggestion|NIT]
```suggestion
	defer cancelStateUpdateSub()
```

File: internal/pkg/agent/application/gateway/fleet/state.go
Comment (AndersonQ) [inline internal/pkg/agent/application/gateway/fleet/state.go:13]: [Nit]
you don't need to say the interface is an interface
```suggestion
// StateFetcher fetches the current state of the coordinator.
```

File: internal/pkg/agent/application/gateway/fleet/scheduler.go
Comment (AndersonQ) [inline internal/pkg/agent/application/gateway/fleet/scheduler.go:10]: [Nit]
you don't need to say the interface is an interface

```suggestion
// Scheduler encapsulate a scheduling logic, this is useful if you want to
// test asynchronous code in a synchronous way.
```

**Review by AndersonQ — APPROVED** [review]

**Comment by cmacknz** [comment]
I think we need to hold off on merging this until we have an easy way for users to switch back to the original 5m timeout without reinstalling the agent. 

This is both to ensure we have a way to go back in case some unforeseen consequence is discovered after upgrade, and also to allow users with network infrastructure they can't easily change to use 8.8 successfully. 

Proxies that terminate the connection after 5 minutes will trigger the agent's retry backoff loop, causing the agent to slowly go offline in the situation where there are no updates from Fleet to short-circuit the long polling connection.

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (pchila) [reply internal/pkg/agent/application/gateway/fleet/fleet_gateway.go:83]: Yes, good catch! :+1:

File: docs/fleet-long-polls.md
Comment (AndersonQ) [inline docs/fleet-long-polls.md:56]:
```suggestion
Due to the way that configuration for fleet-managed agents is handled, it is not currently possible to change such values from whatever the default values were when the agent has been enrolled. For this reason the fleet gateway debounce, interval and backoff settings contained in `FleetGatewaySettings` struct has not been included in the overall `FleetAgentConfig` struct.
```

**Review by AndersonQ — DISMISSED** [review]

**Comment by cmacknz** [comment]
Closing this one for now, we can reopen when we want to revive this.

**Pull request closed without merging**
